src/main/java/com/google/devtools/build/lib/analysis/platform/ToolchainTypeInfo.java

```diff
@@ -0,0 +1,70 @@
+package com.google.devtools.build.lib.analysis.platform;
+
+import com.google.common.base.Strings;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import com.google.devtools.build.lib.events.Location;
+import com.google.devtools.build.lib.packages.NativeInfo;
+import com.google.devtools.build.lib.packages.NativeProvider;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec.VisibleForSerialization;
+import com.google.devtools.build.lib.skylarkbuildapi.platform.ToolchainTypeInfoApi;
+import com.google.devtools.build.lib.skylarkinterface.SkylarkCallable;
+import com.google.devtools.build.lib.skylarkinterface.SkylarkModule;
+import com.google.devtools.build.lib.skylarkinterface.SkylarkModuleCategory;
+import com.google.devtools.build.lib.skylarkinterface.SkylarkPrinter;
+import java.util.Objects;
+
+/** A provider that supplies information about a specific toolchain type. */
+@Immutable
+@AutoCodec
+public class ToolchainTypeInfo extends NativeInfo implements ToolchainTypeInfoApi {
+  /** Name used in Skylark for accessing this provider. */
+  public static final String SKYLARK_NAME = "ToolchainTypeInfo";
+
+  /** Skylark constructor and identifier for this provider. */
+  @AutoCodec
+  public static final NativeProvider<ToolchainTypeInfo> PROVIDER =
+      new NativeProvider<ToolchainTypeInfo>(ToolchainTypeInfo.class, SKYLARK_NAME) {};
+
+  private final Label typeLabel;
```

Comment:
I have a branch in progress that adds a (string-valued) provider_id attribute, which can then be checked against the actual provider during toolchain resolution.

Reply:
Understood! I don't know what your plans are and thus I can't say anything wise about them, but I think this change by itself is a net win, since I consider nodep labels a hack (they are confusing because label attributes _usually_ mean a dependency).

I'm somewhat worried about the "string-valued provider_id attribute" part, but we can discuss that in a separate thread.